crypto/ocsp/ocsp_client.c

```diff
@@ -30,3 +30,98 @@ OCSP_BASICRESP *OCSP_response_get1_basic(OCSP_RESPONSE *resp) {
   return ASN1_item_unpack(rb->response, ASN1_ITEM_rptr(OCSP_BASICRESP));
 }
 
+OCSP_SINGLERESP *OCSP_resp_get0(OCSP_BASICRESP *bs, size_t idx)
+{
+  if (bs == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return NULL;
+  }
+  return sk_OCSP_SINGLERESP_value(bs->tbsResponseData->responses, idx);
+}
+
+int OCSP_resp_find(OCSP_BASICRESP *bs, OCSP_CERTID *id, int last)
+{
+  if (bs == NULL || id == NULL){
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  STACK_OF(OCSP_SINGLERESP) *sresp = bs->tbsResponseData->responses;
+  OCSP_SINGLERESP *single;
+
+  if (last < 0) {
+    last = 0;
+  } else {
+    last++;
+  }
+  for (size_t i = last; i < sk_OCSP_SINGLERESP_num(sresp); i++) {
+    single = sk_OCSP_SINGLERESP_value(sresp, i);
+    if (!OCSP_id_cmp(id, single->certId)) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+int OCSP_single_get0_status(OCSP_SINGLERESP *single, int *reason,
+                            ASN1_GENERALIZEDTIME **revtime,
+                            ASN1_GENERALIZEDTIME **thisupd,
+                            ASN1_GENERALIZEDTIME **nextupd) {
+  if (single == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  OCSP_CERTSTATUS *cst = single->certStatus;
+  int status = cst->type;
+  /*
+   * If certificate status is revoked, we look up certificate revocation
+   * time and reason
+   */
+  if (status == V_OCSP_CERTSTATUS_REVOKED) {
+    OCSP_REVOKEDINFO *rev = cst->value.revoked;
+    if (revtime != NULL) {
+      *revtime = rev->revocationTime;
```

Comment:
Yeah, that makes a lot of sense. This pointer memory assumption was taken over from OpenSSL, but I also think it's safer to copy memory instead of pointing to the same pointer instead. I'll fix it for the returned time values.